modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/ConcurrentStripedPool.java

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright 2019 GridGain Systems, Inc. and Contributors.
+ *
+ * Licensed under the GridGain Community Edition License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.gridgain.com/products/software/community-edition/gridgain-community-edition-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.query.h2;
+
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.LongAdder;
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Concurrent pool of object based on ConcurrentLinkedDeque.
+ */
+public class ConcurrentStripedPool<E> implements Iterable<E> {
+    /** Stripe pools. */
+    private final ConcurrentLinkedQueue<E>[] stripePools;
+
+    /** Stripe pools size (calculates fast, optimistic and approximate). */
+    private LongAdder[] stripeSize;
+
+    /** Stripes count. */
+    private final int stripes;
+
+    /**
+     * Constructor.
+     *
+     * @param stripes Count of stripes.
+     */
+    @SuppressWarnings("unchecked")
+    public ConcurrentStripedPool(int stripes) {
+        this.stripes = stripes;
+
+        stripePools = new ConcurrentLinkedQueue[stripes];
+        stripeSize = new LongAdder[stripes];
+
+        for (int i = 0; i < stripes; ++i) {
+            stripePools[i] = new ConcurrentLinkedQueue<>();
+            stripeSize[i] = new LongAdder();
+        }
+    }
+
+    /**
+     * Pushes an element onto the pool.
+     *
+     * @param e the element to push
+     * @throws NullPointerException if the specified element is null and this
+     *         deque does not permit null elements
+     */
+    public void recycle(E e) {
+        int idx = (int)(Thread.currentThread().getId() % stripes);
+
+        stripePools[idx].add(e);
+
+        stripeSize[idx].increment();
+    }
+
+    /**
+     * Retrieves element from pool, or returns {@code null} if the pool is empty.
+     *
+     * @return the  element of the pool, or {@code null} if the pool is empty.
+     */
+    public E borrow() {
+        int idx = (int)(Thread.currentThread().getId() % stripes);
+
+        E r = stripePools[idx].poll();
+
+        if (r != null)
+         stripeSize[idx].decrement();
+
+        return r;
+    }
+
+    /**
+     * @return Approximate size of pool (faster than ConcurrentLinkedDeque#size()).
+     */
+    public int size() {
```

Comment:
Looks weird. Actually, this is a stripe size.
I'd expect a single LongAdder be used instead of LongAdder[]
or pool should be bounded to some size and all size calculation/borrow/recycle logic should be incapsulated.